src/test/java/org/jnode/fs/xfs/XfsFileSystemTest.java

```diff
@@ -190,12 +191,44 @@ public void testShortFormAttribute() throws Exception {
             assertThat(attributes, hasSize(1));
             FSAttribute attribute = attributes.get(0);
             assertThat(attribute.getName(), is("selinux"));
-            String stringValue = new String(attribute.getValue(), StandardCharsets.UTF_8)
-                    .replace("\0", "");
+            String stringValue = bytes2string(attribute.getValue());
             assertThat(stringValue, is("unconfined_u:object_r:unlabeled_t:s0"));
         }
     }
 
+    @Ignore("test data not in project, it is 17GB, too large to put in code.")
+    @Test
+    public void testBtreeWithHierarchy() throws Exception {
+        File testFile = FileSystemTestUtils.getTestFile("org/jnode/fs/xfs/v5/centos-xfs.img");
+        try (FileDevice device = new FileDevice(testFile, "r")) {
+            XfsEntry entry = DataTestUtils.getDescendantData(new XfsFileSystemType().create(device, true), "usr", "lib64");
+
+            // It's the Btree node with complex folder / file hierarchy.
+            XfsDirectory directory = new XfsDirectory(entry);
+            assertThat(directory.getEntry().getINode().getFormat(), is(Format.BTREE));
+
+            // It seems that all items in the same folder have the same attribute value of name "selinux".
+            FSEntry file = directory.getEntry("librt.so.1"); // a file
+            assertThat(bytes2string(file.getAttributes().get(0).getValue()), is("system_u:object_r:lib_t:s0"));
+
+            FSEntry folder = directory.getEntry("dri"); // a folder
+            assertThat(bytes2string(folder.getAttributes().get(0).getValue()), is("system_u:object_r:lib_t:s0"));
+
+            FSEntry file1InFolder = new XfsDirectory((XfsEntry)folder).getEntry("i965_dri.so");
+            assertThat(bytes2string(file1InFolder.getAttributes().get(0).getValue()), is("system_u:object_r:textrel_shlib_t:s0"));
+
+            FSEntry file2InFolder = new XfsDirectory((XfsEntry)folder).getEntry("r600_dri.so");
+            assertThat(bytes2string(file2InFolder.getAttributes().get(0).getValue()), is("system_u:object_r:textrel_shlib_t:s0"));
+        }
+        finally {
+            testFile.delete();
+        }
+    }
+
+    private String bytes2string(byte[] bytes) {
```

Comment:
Would still be nice to have this as `bytesToString`.
I get the feeling this is also used in production code somewhere so maybe it would be useful to have this in a utility class somewhere?

Reply:
OK, will change it. Thanks.